implement corruption correcting recv

alek-p opened the pull request:
This patch implements a new type of zfs receive: corrective receive (-c). This type of recv is used to heal corrupted data when a replica of the data already exists (in the form of a sendfile for example).
Metadata can not be healed using a corrective receive.

This patch enables us to receive a send stream into an existing snapshot for the purpose of correcting data corruption. 
<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
In the past in the rare cases where ZFS has experienced permanent data corruption, full recovery of the dataset(s) has not always been possible even if replicas existed.
This patch makes recovery from permanent data corruption possible.

### Description
<!--- Describe your changes in detail -->
For every write and spill record in the send stream, we read the corresponding block from disk and if that read fails with a checksum error we overwrite that block with data from the send stream.
After the data is healed we reread the block to make sure it's healed and remove the healed blocks form the corruption lists seen in zpool status.

To makes sure will have correctly matched the data in the send stream to the right dataset to heal there is a restriction that the GUID for the snapshot being received into must match the GUID in the send stream. There are likely several snapshots referring to the same potentially corrupted data so there may be many snapshots with the above condition holding that are able to heal a single block.

The other thing to point out is that we can only correct data. Specifically, we are only able to heal records of type DRR_WRITE and DRR_SPILL since those are the only ones that contain all of the data needed to recreate the damaged block.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
I've been running unit testing very similar to the test that I've added to the zfs-tests

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

alek-p commented:
The next logical extension for part two of this work is to provide a way for a corrupted pool to tell a backup system to generate a minimal send stream in such a way as to enable the corrupted pool to be healed with this generated send stream.
The interface could be something like the following, but maybe there are better suggestions?
```
# dumps spa err list that are part of this snapshot and the snapshot guid
zfs send -C data/fs@snap > /tmp/errlist 

# on replica system generates healing sendfile based on the errors list
zfs send -cc /tmp/errlist backup_data > /tmp/healing_sendfile

# heal our data with the minimal healing sendfile
zfs recv -c data/fs@snap < /tmp/healing_sendfile
```

behlendorf commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +130,103 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+	remove_error_from_list(spa, &spa->spa_errlist_last, zb);
+	remove_error_from_list(spa, &spa->spa_errlist_scrub, zb);
+	spa_add_healed_error(spa, spa->spa_errlog_last, zb);
+	spa_add_healed_error(spa, spa->spa_errlog_scrub, zb);
```
This logic seems overly complicated to me, why couldn't we instead add every healed bookmark to the `spa_errlist_healed` tree and then update `spa_errlog_sync` to update the on-disk errlog accordingly.  This would be very similar to how the existing `spa_errlog_last` is managed.  `spa_errlog_sync()` is already called from `spa_sync_iterate_to_convergence` so you'd want to just make sure it doesn't bail early when `avl_numnodes(&spa->spa_errlist_healed) != 0`.  Or am I missing something?

behlendorf commented on module/zfs/zfs_ioctl.c:
```diff
@@ -5155,6 +5155,7 @@ static const zfs_ioc_key_t zfs_keys_recv_new[] = {
 	{"begin_record",	DATA_TYPE_BYTE_ARRAY,	0},
 	{"input_fd",		DATA_TYPE_INT32,	0},
 	{"force",		DATA_TYPE_BOOLEAN,	ZK_OPTIONAL},
+	{"heal",		DATA_TYPE_BOOLEAN,	ZK_OPTIONAL},
```
You'll want to add this to the comment above and it should be possible to add some test coverage in `zfs-tests/cmd/libzfs_input_check/libzfs_input_check.c`.  At a minimum please document the optional argument (similar to resumable send)

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_corrective.ksh:
```diff
@@ -0,0 +1,135 @@
+{
+	log_must dd bs=512 count=1 if=garbage conv=notrunc \
+	    oflag=sync of=$DEV_RDSKDIR/$DISK seek=$((($1 / 512) + (0x400000 / 512)))
+}
```
Good news, commit b63e2d881f859f0c7d8596be2759dd096e324f48 added support to master to inject targeted damage in to file blocks.  Please use it to ensure this test is reliable.

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_corrective.ksh:
```diff
@@ -0,0 +1,135 @@
+	corrupt_offset "$mid_offset"
+
+	log_must zpool scrub $TESTPOOL
+	log_must sleep 5 # let scrub finish
```
Also added to master is the new `zpool wait` subcommand, you can now run `log_must zpool wait -t scrub $TESTPOOL` and avoid this ugliness.

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_corrective.ksh:
```diff
@@ -0,0 +1,135 @@
+log_assert "ZFS corrective receive should be able to heal corruption"
+
+# we will use this as the source of corruption
+log_must dd if=/dev/urandom of=garbage bs=512 count=1 oflag=sync
```
You should be able to get rid of the garbage file after switching to the targeted injection.

ahrens commented on module/zfs/dmu_recv.c:
```diff
@@ -391,8 +399,21 @@ recv_begin_check_existing_impl(dmu_recv_begin_arg_t *drba, dsl_dataset_t *ds,
+		 * snapshot as the one we are trying to heal.
+		 */
+		struct drr_begin *drrb = drba->drba_cookie->drc_drrb;
+		error = dsl_dataset_hold_obj(dp, val, FTAG, &snap);
```
if `error` is nonzero, shouldn't we be returning that?

ahrens commented on module/zfs/dmu_recv.c:
```diff
@@ -361,12 +365,16 @@ recv_begin_check_existing_impl(dmu_recv_begin_arg_t *drba, dsl_dataset_t *ds,
+	/* New snapshot name must not exist if we're not healing it */
 	error = zap_lookup(dp->dp_meta_objset,
 	    dsl_dataset_phys(ds)->ds_snapnames_zapobj,
 	    drba->drba_cookie->drc_tosnap, 8, 1, &val);
```
Since we're now using `val` as the snapshot's object number, how about renaming it to reflect that, e.g. `snapobj`

ahrens commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +130,103 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+	if (avl_is_empty(&spa->spa_errlist_healed)) {
+		mutex_exit(&spa->spa_errlist_lock);
+		return;
+	}
```
I don't think this `avl_is_empty` case is needed - the right thing will happen below (i.e. the first call to avl_destroy_nodes will return NULL).

ahrens commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +130,103 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+
+/*
+ * If a healed bookmark matches an entry in the error log we stash it in a tree
+ * so that we can later remove the related log entries in sync context.
```
Could there be a race condition where the entry is in the in-memory errlist but not the on-disk errlog, so we don't clear it?  It seems like it would be simpler for this function to always add an entry to the spa_errlist_healed (if it isn't already there), and let spa_remove_healed_errors() handle the case where it isn't actually in the on-disk structure (as it already does).

ahrens commented on module/zfs/dmu_recv.c:
```diff
@@ -2558,6 +2589,173 @@ receive_process_record(struct receive_writer_arg *rwa,
+			goto cleanup;
+		}
+		blkid =
+		    dbuf_whichblock(DB_DNODE((dmu_buf_impl_t *)dbp), 0, offset);
```
You need to use `DB_DNODE_ENTER/EXIT` around the `DB_DNODE()`, or better yet `dmu_buf_dnode_enter()` as you do below.  Or if you're going to cast the dbuf, you might as well just dereference `db_blkid`

ahrens commented on module/zfs/dmu_recv.c:
```diff
@@ -2558,6 +2589,173 @@ receive_process_record(struct receive_writer_arg *rwa,
+		buf = kmem_alloc(lsize, KM_SLEEP);
+		/* Try to read the object to see if it needs healing */
+		err = dmu_read(os, obj, offset, lsize, buf,
+		    DMU_READ_NO_PREFETCH);
```
Why no prefetching?

ahrens commented on module/zfs/dmu_recv.c:
```diff
@@ -2558,6 +2589,173 @@ receive_process_record(struct receive_writer_arg *rwa,
+
+	/* Correct the corruption in place */
+	err = zio_wait(zio_rewrite(NULL, os->os_spa, 0, bp, abd, size, NULL,
+	    NULL, ZIO_PRIORITY_SYNC_WRITE, flags, NULL));
```
Is there any check that the `size` is the same as the BP's psize (possibly rounded up to ashift)?  Or at least that this can't write past the end of what's allocated for the BP?

ahrens commented on module/zfs/dmu_recv.c:
```diff
@@ -2558,6 +2589,173 @@ receive_process_record(struct receive_writer_arg *rwa,
+	if (arc_get_compression(rrd->arc_buf) != BP_GET_COMPRESS(bp)) {
+		/*
+		 * The compression in the stream doesn't match what we had
+		 * on disk; we need to re-compress the buf into the
```
Do we really need to handle this case?  Seems like we could say that it needs to be a `zfs send --compressed` stream to use it with `zfs recv -c`.  Plus, trying to recompress it and get the exact same byte stream adds more restrictions on changing the checksum algorithm implementations, which we'd like to avoid. (cc @allanjude)

ahrens commented on module/zfs/dmu_recv.c:
```diff
@@ -2558,6 +2589,173 @@ receive_process_record(struct receive_writer_arg *rwa,
+
+	/* Correct the corruption in place */
+	err = zio_wait(zio_rewrite(NULL, os->os_spa, 0, bp, abd, size, NULL,
+	    NULL, ZIO_PRIORITY_SYNC_WRITE, flags, NULL));
```
Given the current functionality, we would expect there to be a LOT more records for non-corrupt blocks, than for corrupt blocks.  So I'm even more concerned about the synchronous (and not eligible for predictive prefetch) `dmu_read()`'s to determine if we want to correct this block.

We'd typically (e.g. with >1 vdev) get better performance by simply always asynchronously `zio_rewrite()`ing every record (i.e. without checking if its actually corrupt).  Since you'd have Nrecords async writes vs the current PR's Nrecords sync reads.  Plus the code would be a lot simpler.

I wonder if we should wait for the extensions (to send only the corrupt blocks) are ready, or at least design `recv -c` to work best in that mode.  E.g. by assuming that nearly all records will be for corrupt blocks?

alek-p commented on cmd/zfs/zfs_main.c:
```diff
@@ -4487,7 +4487,7 @@ zfs_do_receive(int argc, char **argv)
 
 	/* check options */
-	while ((c = getopt(argc, argv, ":o:x:dehnuvFsA")) != -1) {
+	while ((c = getopt(argc, argv, ":o:x:dehnuvFsAc")) != -1) {
```
part two of this work will likely need long-ops but I'd like to avoid adding it now

alek-p commented:
Thanks for the reviews guys. I've addressed most of the comments in the version I just pushed. I'm still thinking about the right way to do the rewrite I/O. Perhaps always rewriting (instead of readin first) is the right way to go hmm...

It seems to be too big a limitation to impose saying we have to use `send --compressed` with corrective recv so I'd like to keep the parts that are able to switch compressions if possible.
Perhaps we can deal with the compression algos that don't recompress the same way seperatley from the other compressison types and only make those send streams be compressed.

alek-p commented:
After talking with my colleagues @datto we think it may be too dangerous to rewrite everything that we encounter in the send stream as writing has the potential to do more damage in the cases where corruption is coming from failing HW for example.
We want to prioritize getting the dataset healthy and not necessarily the performance of the healing. In theory, recv healing should be a rare event that does not need to be highly performant. Having said that I'm still working on the way that I/O is done for corrective recv.

The other open question was about the way to handle compression alogos that may not recompress the same block the same way. The suggestion here to avoid this problem we will only heal when the checksum of the data to be used for healing matches the checksum already on disk.

The last thing I'm still working on is to make sure raw and large block send streams are compatible with healing recv.

ahrens commented:
Superseded by #9372
